### ISSUE TITLE
Question about visualization

### ISSUE THREAD

**Issue by jiaruHithub** [rusty1s/pytorch_geometric#1466]
## ❓ Questions & Help

Does PyG have a visual function?And how can I make a graph visual?
If I want to visual a graph in dataset PROTEINS，how should I do?
Thank you!

**Comment by rusty1s** [comment]
In general, we do not provide visualization tools for graphs. However, there are a lot of useful libraries for visualizing graphs, *e.g.* `networkx`. You can use them by converting your data via `torch_geometric.utils.to_networkx`.

**Comment by jiaruHithub** [comment]
Fine,thank you.
What do you think should be targeted in graph pooling?Like the clustering situation or structure changing after pooling?
I have some trouble about it.Thank you very much!

**Comment by rusty1s** [comment]
What do you mean exactly? Are we still talking about visualization?

**Comment by jiaruHithub** [comment]
> What do you mean exactly? Are we still talking about visualization?

Yes, I want to ask you advise about what is necessary visualizations in graph-pooling papers

**Comment by rusty1s** [comment]
Well, that's hard to tell and depends on the task at hand. For example, you can visualize the resulting adjacency matrix, or you can visualize your graph using T-SNE embeddings for nodes. You may want to draw inspiration from a few graph pooling papers, *e.g.*, [this one](https://arxiv.org/pdf/2006.16904.pdf).

**Comment by jiaruHithub** [comment]
Right!Thank you very much!!!
But I can't use it yet.Could you recommend me some The tutorial of visualization(Like use networkx)?
The figure(like fig3 and fig4) in the paper that you told me is so good.Can they all be plotted by networkx?
Thank you.

**Comment by rusty1s** [comment]
I don't have any tutorials for that. I think Figure 3 is plotted via matplotlib, while Figure 4 is plotted via networkx. You can find more about drawing graphs in networkx [here](https://networkx.github.io/documentation/stable/reference/drawing.html).

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
Oh,Thank you,I will study it.

**Comment by jiaruHithub** [comment]
waht tools did he use in GCN paper?(GIF in that paper, reflected the nodes are divided into few clusters)

**Comment by rusty1s** [comment]
That‘s a dimensionality reduction to 2-dimensions using t-SNE on the hidden representations (sklearn has an implementation for that).

**Comment by jiaruHithub** [comment]
> That‘s a dimensionality reduction to 2-dimensions using t-SNE on the hidden representations (sklearn has an implementation for that).

Oh,Thank you,after use the t-SEN,Output one picture by networkx once per training batch,then make them a GIF.
Is that right?Or GIF is a one-off?

**Comment by rusty1s** [comment]
I‘m not sure what you mean by GIF. I was referring to Figure 1b, which simply plots the 2-dimensional points without any graph structure involved.

**Comment by jiaruHithub** [comment]
> I‘m not sure what you mean by GIF. I was referring to Figure 1b, which simply plots the 2-dimensional points without any graph structure involved.

Oh, fine.I was referring to Figure 4(a~f).And kipf made a video(Figure4 are the some screenshots from the video).
Figure 4 reflected the the process of nodes classififcition(or clustering).It made a deep impression for me.

**Comment by rusty1s** [comment]
Ok, then you simply need to plot that for every epoch :)

**Comment by jiaruHithub** [comment]
> Ok, then you simply need to plot that for every epoch :)

Oh,thank you!But I also have some questions about understanding clustering.
After graph through the GCN and MLP,it can make the nodes have their own label(wong or right).
But, why it can reflected the position changing in figure ,which make those nodes look like they are clustering.

**Comment by rusty1s** [comment]
The clustering is based on the hidden representations, which should ideally cluster for nodes with the same labels.

**Comment by jiaruHithub** [comment]
> The plotting is based on the hidden representations, which should ideally cluster for nodes with the same labels.
Are you saying that the plotting system of networkx will automatically clustering the nodes have same labels?

**Comment by rusty1s** [comment]
No, only if you use positions for nodes based on their hidden representations.

**Comment by jiaruHithub** [comment]
> No, only if you use positions for nodes based on their hidden representations.

OH!Position  in figure is depend on velue of hidden representation.
After GCN layers ,the hidden representation of nodes have same labels will more similar.So the position of nodes(have same label) is more recent.
Am I right?

**Comment by rusty1s** [comment]
Right :)

**Comment by jiaruHithub** [comment]
Thank you,friends.
I didn't realize we'd been communicating for so long.
I wish you best life.
